src/PrintReceipt.ts

```diff
@@ -1,5 +1,121 @@
 import {loadAllItems, loadPromotions} from './Dependencies'
 
+interface ItemDto 
+{
+  barcode: string,
+  name: string,
+  unit: string,
+  price: number
+}
+interface PromotionDto
+{
+  type: string,
+  barcodes: string[]
+}
+
+export interface ReceiptItemDto 
+{
+  barcode: string,
+  quantity: number,
+  totalPrice: number,
+}
+
 export function printReceipt(tags: string[]): string {
-  return ''
+  const items: ItemDto[] = loadAllItems();
+  const promotions: PromotionDto[] = loadPromotions();
+
+  const receiptItems : ReceiptItemDto[] = recordQuantityAndCalcPrice(tags, items);
+  
+  const discount = applyPromotion(receiptItems, promotions);
+
+  const resultText = render(receiptItems, items, discount);
+
+  return resultText;
 }
+
+export function render(receiptItems: ReceiptItemDto[], allItems: ItemDto[], discount: number): string
+{  
+  let resText = `***<store earning no money>Receipt ***\n`;
+  let totalPrice = 0;
+  for (let receiptItem of receiptItems)
+  {
+    const itemDescription = allItems.find(i => i.barcode === receiptItem.barcode)
+    resText += `Name：${itemDescription?.name}，`;
+    resText += `Quantity：${receiptItem.quantity} ${itemDescription?.unit}s，`;
+    resText += `Unit：${itemDescription?.price.toFixed(2)}(yuan)，`;
+    resText += `Subtotal：${receiptItem.totalPrice.toFixed(2)}(yuan)\n`;
+
+    totalPrice += receiptItem.totalPrice;
+  }
+  resText += "----------------------\n";
+  resText += `Total：${totalPrice.toFixed(2)}(yuan)\n`
+  resText += `Discounted prices：${discount.toFixed(2)}(yuan)\n`
+  resText += `**********************`;
+  
+  return resText;
+}
+
+export function applyPromotion(receiptItems: ReceiptItemDto[], promotions: PromotionDto[]): number {
+  let allDiscount = 0;
+  const promotion = promotions[0];
+  for (let item of receiptItems)
+  {
+    if (promotion.barcodes.includes(item.barcode))
+    {
+      const freeAmount = Math.floor(item.quantity / 3);
+      const discount = item.totalPrice / item.quantity * freeAmount;
+      item.totalPrice = item.totalPrice - discount;
+      allDiscount += discount;
+    }
+  }
+  return allDiscount;
+}
+
+export function recordQuantityAndCalcPrice(tags: string[], items: ItemDto[]): ReceiptItemDto[]
+{
+  const res: ReceiptItemDto[] = [];
+  for(let tag of tags)
+  {
+    let deserializeTag = deserializeTagToReceiptItem(tag);
```

Comment:
deserializeTag looks like a tag, not a DTO

Reply:
parseAToB may be a good name

Reply:
Agree